Open new SMB Connection crashes

**Describe the bug**
Tried to open up a SMB connection, but crash happens when clicking on the option.

I'm on [grapheneOS](https://grapheneos.org/), where permissons might be stricter.
If testes it with all availabe permissions enabled though, which would be 'Network', 'Notifications', 'Sensors'.

**To Reproduce**
Steps to reproduce the behavior:
1. Click on '+'
2. Click on 'Cloud Connection'
3. Select 'SMB Connection'
4. See crash log

## Exception
* __App Name:__ Amaze File Manager
* __Package:__ com.amaze.filemanager
* __Version:__ 3.10
* __User Action:__ UI Error
* __Request:__ Application crash
* __OS:__ Linux Android 15 - 35
* __Device:__ akita
* __Model:__ Pixel 8a
* __Product:__ akita
<details><summary><b>Crash log </b></summary><p>

```
io.reactivex.exceptions.OnErrorNotImplementedException: The exception was not handled due to missing onError handler in the subscribe() method call. Further reading: https://github.com/ReactiveX/RxJava/wiki/Error-Handling | java.io.IOException: sendto failed: EPERM (Operation not permitted)
	at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:704)
	at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:701)
	at io.reactivex.internal.observers.LambdaObserver.onError(LambdaObserver.java:77)
	at io.reactivex.internal.operators.observable.ObservableDoOnEach$DoOnEachObserver.onError(ObservableDoOnEach.java:117)
	at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.checkTerminated(ObservableObserveOn.java:281)
	at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:172)
	at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:255)
	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:307)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
	at java.lang.Thread.run(Thread.java:1012)
Caused by: java.io.IOException: sendto failed: EPERM (Operation not permitted)
	at libcore.io.IoBridge.maybeThrowAfterSendto(IoBridge.java:721)
	at libcore.io.IoBridge.sendto(IoBridge.java:688)
	at java.net.PlainDatagramSocketImpl.send(PlainDatagramSocketImpl.java:126)
	at java.net.DatagramSocket.send(DatagramSocket.java:723)
	at com.amaze.filemanager.utils.smb.WsddDiscoverDeviceStrategy.multicastForDevice(WsddDiscoverDeviceStrategy.kt:128)
	at com.amaze.filemanager.utils.smb.WsddDiscoverDeviceStrategy.discoverDevices(WsddDiscoverDeviceStrategy.kt:96)
	at com.amaze.filemanager.utils.smb.SmbDeviceScannerObservable.subscribeActual$lambda$1$lambda$0(SmbDeviceScannerObservable.kt:89)
	at com.amaze.filemanager.utils.smb.SmbDeviceScannerObservable.$r8$lambda$3Q3UXszuikPQxo-e_j1qIR2is2M(Unknown Source:0)
	at com.amaze.filemanager.utils.smb.SmbDeviceScannerObservable$$ExternalSyntheticLambda0.call(Unknown Source:4)
	at io.reactivex.internal.operators.observable.ObservableFromCallable.subscribeActual(ObservableFromCallable.java:43)
	at io.reactivex.Observable.subscribe(Observable.java:12267)
	at io.reactivex.internal.operators.observable.ObservableSubscribeOn$SubscribeTask.run(ObservableSubscribeOn.java:96)
	at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
	... 7 more
Caused by: android.system.ErrnoException: sendto failed: EPERM (Operation not permitted)
	at libcore.io.Linux.sendtoBytes(Native Method)
	at libcore.io.Linux.sendto(Linux.java:235)
	at libcore.io.ForwardingOs.sendto(ForwardingOs.java:718)
	at libcore.io.BlockGuardOs.sendto(BlockGuardOs.java:398)
	at libcore.io.ForwardingOs.sendto(ForwardingOs.java:718)
	at libcore.io.IoBridge.sendto(IoBridge.java:686)
	... 18 more

```
</details>
<hr>

--- a/app/src/main/java/com/amaze/filemanager/ui/notifications/FtpNotification.java
+++ b/app/src/main/java/com/amaze/filemanager/ui/notifications/FtpNotification.java
@@ -103,7 +103,7 @@ public class FtpNotification {
     boolean secureConnection =
         sharedPreferences.getBoolean(FtpService.KEY_PREFERENCE_SECURE, FtpService.DEFAULT_SECURE);
 
-    InetAddress address = NetworkUtil.getLocalInetAddress(context);
+    InetAddress address = NetworkUtil.getLocalInetAddress(context, false);
 
     String address_text = "Address not found";
 

--- a/app/src/main/java/com/amaze/filemanager/utils/NetworkUtil.kt
+++ b/app/src/main/java/com/amaze/filemanager/utils/NetworkUtil.kt
@@ -100,7 +100,10 @@ object NetworkUtil {
      * Caveat: doesn't handle IPv6 addresses well. Forcing return IPv4 if possible.
      */
     @JvmStatic
-    fun getLocalInetAddress(context: Context): InetAddress? {
+    fun getLocalInetAddress(
+        context: Context,
+        requestMulticast: Boolean = false,
+    ): InetAddress? {
         if (!isConnectedToLocalNetwork(context)) {
             return null
         }
@@ -112,14 +115,25 @@ object NetworkUtil {
             return if (ipAddress == 0) null else intToInet(ipAddress)
         }
         runCatching {
-            NetworkInterface.getNetworkInterfaces().iterator().forEach { netinterface ->
-                netinterface.inetAddresses.iterator().forEach { address ->
+            NetworkInterface.getNetworkInterfaces().iterator().forEach { networkInterface ->
+                networkInterface.inetAddresses.iterator().forEach { address ->
                     // this is the condition that sometimes gives problems
                     if (!address.isLoopbackAddress &&
                         !address.isLinkLocalAddress &&
                         address is Inet4Address
                     ) {
-                        return address
+                        if (requestMulticast) {
+                            if (networkInterface.supportsMulticast()) {
+                                return address
+                            } else {
+                                log.warn(
+                                    "network interface {} does not support multicast",
+                                    networkInterface.displayName,
+                                )
+                            }
+                        } else {
+                            return address
+                        }
                     }
                 }
             }

--- a/app/src/main/java/com/amaze/filemanager/utils/smb/SameSubnetDiscoverDeviceStrategy.kt
+++ b/app/src/main/java/com/amaze/filemanager/utils/smb/SameSubnetDiscoverDeviceStrategy.kt
@@ -95,7 +95,7 @@ class SameSubnetDiscoverDeviceStrategy : SmbDeviceScannerObservable.DiscoverDevi
     }
 
     private fun getNeighbourhoodHosts(): List<InetAddress> {
-        val deviceAddress = NetworkUtil.getLocalInetAddress(AppConfig.getInstance())
+        val deviceAddress = NetworkUtil.getLocalInetAddress(AppConfig.getInstance(), requestMulticast = true)
         return deviceAddress?.let { addr ->
             if (addr is Inet6Address) {
                 // IPv6 neigbourhood hosts can be very big - that should use wsdd instead; hence

--- a/app/src/main/java/com/amaze/filemanager/utils/smb/WsddDiscoverDeviceStrategy.kt
+++ b/app/src/main/java/com/amaze/filemanager/utils/smb/WsddDiscoverDeviceStrategy.kt
@@ -103,7 +103,7 @@ class WsddDiscoverDeviceStrategy : SmbDeviceScannerObservable.DiscoverDeviceStra
 
     @Suppress("LabeledExpression")
     private fun multicastForDevice(callback: (ComputerParcelable) -> Unit) {
-        NetworkUtil.getLocalInetAddress(AppConfig.getInstance())?.let { addr ->
+        NetworkUtil.getLocalInetAddress(AppConfig.getInstance(), requestMulticast = true)?.let { addr ->
             val multicastAddressV4 = InetAddress.getByName(BROADCAST_IPV4)
             val multicastAddressV6 = InetAddress.getByName(BROADCAST_IPV6_LINK_LOCAL)
 

--- a/app/src/test/java/com/amaze/filemanager/utils/NetworkUtilTest.kt
+++ b/app/src/test/java/com/amaze/filemanager/utils/NetworkUtilTest.kt
@@ -1,0 +1,267 @@
+package com.amaze.filemanager.utils
+
+import android.app.Service
+import android.content.Context
+import android.net.ConnectivityManager
+import android.net.Network
+import android.net.NetworkCapabilities
+import android.net.NetworkInfo
+import android.net.wifi.WifiInfo
+import android.net.wifi.WifiManager
+import android.os.Build
+import android.os.Build.VERSION_CODES
+import android.os.Build.VERSION_CODES.LOLLIPOP
+import android.os.Build.VERSION_CODES.P
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.mockkStatic
+import io.mockk.unmockkAll
+import org.junit.After
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNotNull
+import org.junit.Assert.assertNull
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.annotation.Config
+import java.net.Inet4Address
+import java.net.NetworkInterface
+import java.util.Collections
+
+/**
+ * Tests for [NetworkUtil].
+ */
+@RunWith(AndroidJUnit4::class)
+@Config(sdk = [LOLLIPOP, P, VERSION_CODES.R])
+@Suppress("StringLiteralDuplication")
+class NetworkUtilTest {
+    private lateinit var context: Context
+    private lateinit var connectivityManager: ConnectivityManager
+    private lateinit var wifiManager: WifiManager
+    private lateinit var wifiInfo: WifiInfo
+
+    /**
+     * Set up the mocks.
+     */
+    @Before
+    fun setUp() {
+        context = mockk(relaxed = true)
+        connectivityManager = mockk(relaxed = true)
+        wifiManager = mockk(relaxed = true)
+        wifiInfo = mockk(relaxed = true)
+
+        every { context.applicationContext } returns context
+        every {
+            context.getSystemService(Service.CONNECTIVITY_SERVICE)
+        } returns connectivityManager
+        every {
+            context.getSystemService(Service.WIFI_SERVICE)
+        } returns wifiManager
+        every { wifiManager.connectionInfo } returns wifiInfo
+    }
+
+    /**
+     * Test [NetworkUtil.isConnectedToLocalNetwork] when device is not connected.
+     */
+    @Test
+    fun testGetLocalInetAddressWhenNotConnected() {
+        mockNetworkNotConnected()
+        assertNull(NetworkUtil.getLocalInetAddress(context))
+    }
+
+    /**
+     * Test [NetworkUtil.getLocalInetAddress] when device is connected to Wifi.
+     */
+    @Test
+    fun testGetLocalInetAddressOnWifi() {
+        mockWifiConnected()
+        every { wifiInfo.ipAddress } returns 0x0F02000A // 10.0.2.15
+
+        val result = NetworkUtil.getLocalInetAddress(context)
+        assertNotNull(result)
+        assertEquals("10.0.2.15", result?.hostAddress)
+    }
+
+    /**
+     * Test [NetworkUtil.getLocalInetAddress] when device is connected to Ethernet.
+     */
+    @Test
+    fun testGetLocalInetAddressOnEthernet() {
+        mockEthernetConnected()
+        mockkStatic(NetworkInterface::class)
+
+        val inetAddress = mockk<Inet4Address>()
+        every { inetAddress.isLoopbackAddress } returns false
+        every { inetAddress.isLinkLocalAddress } returns false
+        every { inetAddress.hostAddress } returns "192.168.1.100"
+
+        val networkInterface = mockk<NetworkInterface>()
+        every { networkInterface.inetAddresses } returns
+            Collections.enumeration(listOf(inetAddress))
+        every { NetworkInterface.getNetworkInterfaces() } returns
+            Collections.enumeration(listOf(networkInterface))
+
+        val result = NetworkUtil.getLocalInetAddress(context)
+        assertNotNull(result)
+        assertEquals("192.168.1.100", result?.hostAddress)
+    }
+
+    /**
+     * Test [NetworkUtil.getLocalInetAddress] for network interface that supports multicast.
+     */
+    @Test
+    fun testGetLocalInetAddressWithMulticastSupport() {
+        mockEthernetConnected()
+        mockkStatic(NetworkInterface::class)
+
+        val inetAddress = mockk<Inet4Address>()
+        every { inetAddress.isLoopbackAddress } returns false
+        every { inetAddress.isLinkLocalAddress } returns false
+        every { inetAddress.hostAddress } returns "192.168.1.1"
+
+        val networkInterface = mockk<NetworkInterface>()
+        every { networkInterface.supportsMulticast() } returns true
+        every { networkInterface.inetAddresses } returns
+            Collections.enumeration(listOf(inetAddress))
+        every { NetworkInterface.getNetworkInterfaces() } returns
+            Collections.enumeration(listOf(networkInterface))
+
+        val result = NetworkUtil.getLocalInetAddress(context, requestMulticast = true)
+        assertNotNull(result)
+        assertEquals("192.168.1.1", result?.hostAddress)
+    }
+
+    /**
+     *
+     */
+    @Test
+    fun testGetLocalInetAddressWithoutMulticastSupport() {
+        mockEthernetConnected()
+
+        val inetAddress = mockk<Inet4Address>()
+        every { inetAddress.isLoopbackAddress } returns false
+        every { inetAddress.isLinkLocalAddress } returns false
+        every { inetAddress.hostAddress } returns "192.168.1.100"
+
+        val networkInterface = mockk<NetworkInterface>()
+        every { networkInterface.supportsMulticast() } returns false
+        every { networkInterface.displayName } returns "eth0"
+        every { networkInterface.inetAddresses } returns
+            Collections.enumeration(listOf(inetAddress))
+
+        mockkStatic(NetworkInterface::class)
+        every { NetworkInterface.getNetworkInterfaces() } returns
+            Collections.enumeration(listOf(networkInterface))
+
+        val result = NetworkUtil.getLocalInetAddress(context, requestMulticast = true)
+        assertNull(result)
+    }
+
+    /**
+     * Test [NetworkUtil.getLocalInetAddress] when there are multiple network interfaces, and we
+     * need to pick one that supports multicast.
+     */
+    @Test
+    fun testGetLocalInetAddressWithMultipleInterfacesOneSupportsMulticast() {
+        mockEthernetConnected()
+        mockkStatic(NetworkInterface::class)
+
+        // Create a non-multicast interface
+        val nonMulticastAddress = mockk<Inet4Address>()
+        every { nonMulticastAddress.isLoopbackAddress } returns false
+        every { nonMulticastAddress.isLinkLocalAddress } returns false
+        every { nonMulticastAddress.hostAddress } returns "192.168.1.100"
+
+        val nonMulticastInterface = mockk<NetworkInterface>()
+        every { nonMulticastInterface.supportsMulticast() } returns false
+        every { nonMulticastInterface.displayName } returns "eth0"
+        every { nonMulticastInterface.inetAddresses } returns
+            Collections.enumeration(listOf(nonMulticastAddress))
+
+        // Create a multicast-capable interface
+        val multicastAddress = mockk<Inet4Address>()
+        every { multicastAddress.isLoopbackAddress } returns false
+        every { multicastAddress.isLinkLocalAddress } returns false
+        every { multicastAddress.hostAddress } returns "192.168.2.100"
+
+        val multicastInterface = mockk<NetworkInterface>()
+        every { multicastInterface.supportsMulticast() } returns true
+        every { multicastInterface.displayName } returns "wlan0"
+        every { multicastInterface.inetAddresses } returns
+            Collections.enumeration(listOf(multicastAddress))
+
+        // Mock NetworkInterface.getNetworkInterfaces() to return both interfaces
+        every { NetworkInterface.getNetworkInterfaces() } returns
+            Collections.enumeration(listOf(nonMulticastInterface, multicastInterface))
+
+        val result = NetworkUtil.getLocalInetAddress(context, requestMulticast = true)
+        assertNotNull(result)
+        assertEquals("192.168.2.100", result?.hostAddress)
+    }
+
+    private fun mockNetworkNotConnected() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            val networkCapabilities = mockk<NetworkCapabilities>()
+            every { networkCapabilities.hasTransport(any()) } returns false
+            every { connectivityManager.activeNetwork } returns null
+            every {
+                connectivityManager.getNetworkCapabilities(any())
+            } returns networkCapabilities
+        } else {
+            every { connectivityManager.activeNetworkInfo } returns null
+        }
+    }
+
+    private fun mockWifiConnected() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            val network = mockk<Network>()
+            val networkCapabilities = mockk<NetworkCapabilities>()
+            every {
+                networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)
+            } returns true
+            every {
+                networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)
+            } returns false
+            every { connectivityManager.activeNetwork } returns network
+            every {
+                connectivityManager.getNetworkCapabilities(network)
+            } returns networkCapabilities
+        } else {
+            val networkInfo = mockk<NetworkInfo>()
+            every { networkInfo.isConnected } returns true
+            every { networkInfo.type } returns ConnectivityManager.TYPE_WIFI
+            every { connectivityManager.activeNetworkInfo } returns networkInfo
+        }
+    }
+
+    private fun mockEthernetConnected() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            val network = mockk<Network>()
+            val networkCapabilities = mockk<NetworkCapabilities>()
+            every {
+                networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)
+            } returns false
+            every {
+                networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)
+            } returns true
+            every { connectivityManager.activeNetwork } returns network
+            every {
+                connectivityManager.getNetworkCapabilities(network)
+            } returns networkCapabilities
+        } else {
+            val networkInfo = mockk<NetworkInfo>()
+            every { networkInfo.isConnected } returns true
+            every { networkInfo.type } returns ConnectivityManager.TYPE_ETHERNET
+            every { connectivityManager.activeNetworkInfo } returns networkInfo
+        }
+    }
+
+    /**
+     * Clean up the mocks.
+     */
+    @After
+    fun tearDown() {
+        unmockkAll()
+    }
+}

